cpp/src/io/avro/reader_impl.cu

```diff
@@ -367,9 +367,20 @@ table_with_metadata reader::impl::read(avro_reader_options const &options,
     }
 
     if (_metadata->total_data_size > 0) {
-      const auto buffer =
-        _source->host_read(_metadata->block_list[0].offset, _metadata->total_data_size);
-      rmm::device_buffer block_data(buffer->data(), buffer->size(), stream);
+      rmm::device_buffer block_data;
+      if (_source->is_device_read_preferred(_metadata->total_data_size)) {
+        rmm::device_buffer data(_metadata->total_data_size, stream);
+        auto read_size = _source->device_read(_metadata->block_list[0].offset,
```

Comment:
I think read_size being different from total_data_size would be an error for the Avro reader. Might be better to just CUDF_EXPECTS here instead of resizing the buffer below.